### DISCUSSION TITLE
Cannot read JSON file from server action when deployed to Vercel.

### DISCUSSION THREAD

**Discussion by Tomekmularczyk** [vercel/next.js#58512] [Help]
### Summary

I tried reading a JSON file from a Server Component like explained in [this blog post](https://vercel.com/guides/loading-static-file-nextjs-api-route) - It works great. 

My problem is that I want to read the same JSON file from a server action, which works on localhost but doesn't work when deployed to Vercel. I get an error `ENOENT: no such file or directory`.

My server action file is very simple:

```tsx
// search.ts

"use server"

export async function getResults(
  _: any[],
  formData: FormData
): Promise<GetResultsResponse> {
  const datas = JSON.parse(
    await fs.readFile(
      process.cwd() + "/src/app/indexes/data.json",
      "utf8"
    )
  );
  console.log(datas);

  return {
     ...
  };
}

```

**Comment by BeSketens** [comment]
I get exactly the same error..

**Comment by patrickn2** [comment]
Same error here

ENOENT: no such file or directory, open '/var/task/app/contact_email.html'

`const emailTemplate = fs.readFileSync(
      process.cwd() + "/app/contact_email.html",
      "utf8"
    );`

**Comment by MinsterRobin** [comment]
Same issue here, any update on this ?

**Comment by VAR0S** [comment]
Same here

**Comment by LeweProjects** [comment]
This fixed my issue, try using API axios to fetch data instead.

in my case I was using NextJS so my project tree would look like this
app
![image](https://github.com/vercel/next.js/assets/120403809/e706e0c3-cff4-4e43-8dce-2c4ef8240428)

```typescript
route.ts:
import fs from "fs";
import path from "path";

export function GET(request: any) {
	let usersPath = path.join(process.cwd(), "/src/app/indexes/data.json");
	let file = fs.readFileSync(usersPath);
	return new Response(file);
}
```


your fetch function should be like this:

```typescript
import { useEffect } from "react";
import axios from "axios";

const useJsonMapping = () => {
        useEffect(() => {
          axios
            .get("/api/fetchProjects")
            .then((res) => console.log(res.data))
            .catch((res) => console.error(res));
	}, []);
}

export default useJsonMapping;  
```

**Comment by noobling** [comment]
Any updates on this? would be good to clarify since its an official blog post

**Comment by kristojorg** [comment]
There are even a second incorrect documentation page here: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

Vercel started out simple but is becoming more like AWS every day :/

**Comment by davidpangsw** [comment]
Meet a similar problem. Not solved yet.
My client-side component invoke an utility function from a source marked as server side:
```
  useEffect(() => {
    (async () => {
      const props = await fetchGraphProps();
...
      setGraphProps(props);
    })();
  }, []);
```
`fetchGraphProps` is from the source marked as server side.

The only work-around I can think of: move the file being read to `public` or `api` and use `fetch`. I haven't tried it though.

**Reply by davidpangsw** [reply]
I move the file to `/public` and:
```
  useEffect(() => {
    (async 
      const res = await fetch(PROPS_PATH);
      const props = await res.json();
...
```
It works. Still looking forward to any better solutions :)
